DOCS-3613: Iterate on Connect Devices (#4037)
viamrobotics/docs: docs/operate/get-started/basics.md

> ---

Reviewer: hey as we chatted about last week - i just want to confirm this is intended to be helpful but not required to get started?  under the assumption that a lot of people very well might skip this page.  Also, i know this might not matter a huge amount, but I think i like renaming this to something without a verb since it is not procedural content.  since github uses 'about github and git' maybe 'About Viam'?  'Overview'?  etc.

Reviewer 2: I actually kind of like the verb - feel like it encourages people to start here :P

Author: Oh I'd thought you wanted verbs across the board, so we came up with this and I somewhat agree with Esha though don't feel strongly. We used to have a page called "About Viam" and some people thought it meant "About Viam as a company" instead of "About the Viam platform." I do think Overview would be okay, and people would be less likely to read it.

If we *want* most people to skip over this page, then we can make it Overview, but then we need to move "What is a machine?" to each of the setup pages because we've gotten overwhelming feedback that that's a key helpful section.

Author: Discussed w/Esha offline and we feel that "Learn the basics" is best here. Yes people can skip over it but with this title, they're more likely to come here if they have questions as they move through other content.

Reviewer: i swear I won't abuse my veto power too much but I do actually feel kind of strongly about keeping verb+object for procedural content and nouns/adjectives/etc (e.g. "About The Viam Platform" or "Viam Overview") for context-setting and/or reference content.

Author: For thread coherency: Ian said via email reply (see conversation page on this PR) fine to leave as-is.

Okay I will change to Viam basics in left nav.

linkTitle: "Learn the basics"
title: "Viam basics"
weight: 10
layout: "docs"
type: "docs"
images: ["/installation/thumbnails/install.png"]
imageAlt: "Install Viam"
no_list: false
description: "Learn the basics of Viam."
carouselscript: true
---

## What is Viam?

Viam is a software platform for data, AI, and automation.

Viam consists of:

- An open-source binary called `viam-server` that runs on your machine, managing your local hardware, software, and data, and connecting your device to Viam's cloud.
- A cloud app called the [Viam app](https://app.viam.com) that you can use to configure and manage your machines and data.
- Simple APIs for common types of hardware (for example, cameras, sensors, and motors), and software services (such as computer vision).
- A registry of modules that implement the hardware and software APIs.
  This includes support for many popular hardware models.

> - SDKs for many programming languages that you can use to control your machines.

Reviewer 2: ```suggestion
- SDKs for many programming languages that you can use to interact with your machines.
```

{{% relatedcard link="/dev/reference/apis/components/encoder/" %}}
{{% relatedcard link="/dev/reference/apis/components/gantry/" %}}
{{% relatedcard link="/dev/reference/apis/components/gripper/" %}}
{{% relatedcard link="/dev/reference/apis/components/input-controller/" %}}
{{% relatedcard link="/dev/reference/apis/components/motor/" %}}
{{% relatedcard link="/dev/reference/apis/components/movement-sensor/" %}}
{{% relatedcard link="/dev/reference/apis/components/power-sensor/" %}}
{{% relatedcard link="/dev/reference/apis/components/sensor/" %}}
{{% relatedcard link="/dev/reference/apis/components/servo/" %}}
{{% relatedcard link="/dev/reference/apis/components/switch/" %}}
{{% relatedcard link="/dev/reference/apis/components/generic/" %}}
{{< /cards >}}

These standardized APIs are implemented by {{< glossary_tooltip term_id="module" text="modules" >}} that provide drivers for specific models of hardware.

Any hardware that is not already supported by a Viam module can be added into Viam's system of modular resources by [creating a new module](/operate/get-started/other-hardware/) that provides a driver for the hardware.

> ## How do I get started?

Author: Okay yeah--removed links and shortened slightly

> 1. [Set up your machine and connect it to the Viam app](/operate/get-started/setup/).

Reviewer 2: One final thought here: if we phrase this as "Set up a computer or SBC or set up an ESP32 and connect it to the Viam app", we could put the two respective links on the text that matches the left nav titles

Author: ```suggestion
1. [Set up a computer or SBC](/operate/get-started/setup/) or [set up an ESP32](/operate/get-started/setup-micro/) and connect it to the Viam app.
```

1. Next, [configure supported hardware](/operate/get-started/supported-hardware/) and software on your machine in the Viam app.

> 1. To integrate other hardware or software services that aren't already supported, you can [write your own modules](/operate/get-started/other-hardware/).

Author: @EshaMaharishi for this one I'm not sure putting the link on the matching text makes as much sense since I want to explain before they click something, and I think keeping the why before the how makes sense.

Reviewer 2: Personally I think it's okay to put the link on the matching text even if it's at the beginning of the sentence, because I think the visual cue that the text matches the link content is pretty powerful. But if you feel strongly, feel free to leave as-is (feel free to resolve)

Author: I guess esp. if we can change the set up one it is worth changing here
```suggestion
1. To [integrate other hardware](/operate/get-started/other-hardware/) or software services that aren't already supported, you can write your own modules.
```

Reviewer 2: Nice :)

---
linkTitle: "Learn the basics"
title: "Viam basics"
weight: 10
layout: "docs"
type: "docs"
images: ["/installation/thumbnails/install.png"]
imageAlt: "Install Viam"
no_list: false
description: "Learn the basics of Viam."
carouselscript: true
---

## What is Viam?

> Viam is a software platform for data, AI, and automation.

Author: If we say "cloud-based" I worry people will immediately assume everything is in the cloud and be rather confused when we try to explain viam-server. But meanwhile, if we just say "platform" sometimes people think it's a hardware platform like a robot base (Viam rover). Though, with "application" right after maybe it'll be fine. Does "cloud-based" need to be here versus just
```suggestion
Viam is a platform that provides the tools and cloud infrastructure to create intelligent physical systems.
```

> Viam is a software platform for data, AI, and automation.

Author: I worry that "smart applications" on its own focuses too much on applications though and not on data storage, fleet management, or hardware support. I know we go into more detail below though so maybe
```suggestion
Viam is a full-stack software platform for building smart applications for the physical world.
```
is okay?

> Viam is a software platform for data, AI, and automation.

Reviewer 2: Hmm. I don't think we use "full-stack" anywhere else...

I would suggest just saying "software platform for building smart applications for the physical world" for now, since it's relatively simple, and let the rest of the content speak for itself. We could always revisit in the future!

Feel free to pick one of the options and resolve

> It drives your hardware (for example, cameras, sensors, and motors) and connects your devices to the cloud.

Reviewer 2: ```suggestion
```

Suggested folding into the bullet about API's below

Viam consists of:

> - An open-source binary called `viam-server` that runs on your machine, managing your hardware, software, and data.

Reviewer 2: ```suggestion
- An open-source binary called `viam-server` that runs on your machine, managing your local hardware and software and connection to Viam's cloud.
```

Author: ```suggestion
- An open-source binary called `viam-server` that runs on your machine, managing your local hardware, software, and data, and connecting your device to Viam's cloud.
```

- A cloud app called the [Viam app](https://app.viam.com) that you can use to configure and manage your machines and data.

> - A registry of modules that support many different types of hardware and software.
> - Standardized APIs that abstract the underlying hardware and software.

Reviewer 2: ```suggestion
- Simple APIs for common types of hardware (for example, cameras, sensors, and motors), and software services (such as computer vision and motion planning)
- A registry of modules that implement the hardware and software API's, including for many popular hardware models
```

- SDKs for many programming languages that you can use to control your machines.

`viam-server` connects your machine to the Viam app and to SDK clients.

See [Viam architecture](/operate/reference/architecture/) for details on how Viam works.

## What is a machine?

A _machine_ is a computer (often a single-board computer like a Raspberry Pi or Jetson) or microcontroller and all the hardware attached to it, as well as the software running on it.

> You can think of one machine as representing one device, or one robot.

Reviewer 2: ```suggestion
```

Suggesting to remove to avoid using "robot", and at that point maybe it's not adding much to say you can think of a machine as representing one device?

Author: I think this is the most concrete part of this though? Like the most ELIF sentence in this section, and people keep saying "what is a machine" was really helpful. I fear that "device" doesn't necessarily conjure the image of multiple connected bits, whereas "robot" does. But I know not all devices are robots. How about this:
```suggestion
You can think of one machine as representing one device, or one discrete physical system such as one robot.
```

Each machine runs an instance of `viam-server`.

When you create a new machine in the Viam app, Viam generates a unique set of credentials for that machine that connect the physical machine to its instance in the Viam app.

## What platforms does Viam run on?

Viam can run on any computer that runs one of the following operating systems:

- Linux 64-bit operating systems running on AArch64 (ARM64) or x86-64 architectures
- macOS
- Windows

> - 32-bit microcontrollers (limited support)

Reviewer 2: ```suggestion

Viam can also run on 32-bit microcontrollers such as the ESP32 series.
```

Suggesting this since above we say "computer" and "operating systems"

Author: Good point

> Examples of computers that can run Viam:

Reviewer 2: ```suggestion
Examples of computing devices that can run Viam:
```

Maybe?

{{< board-carousel >}}

## What hardware does Viam support?

> In short, anything.

Reviewer 2: ```suggestion
```

Maybe a little sassy :P

I think this is already covered by "Any hardware that is not already supported..." below

Author: Yeah this was experimental haha. It's so to-the-point though! I do want this to be super clear without someone needing to do the mental work of "Hmm okay so they support arms and motors and grippers and cameras...scanning....do they support sensors?...oh yes okay...what don't they support? maybe nothing? What about microphones? Is there an API for that? Now I have to read about generic components and modules and such"

How about this:

```suggestion
In short, essentially any type of hardware can be integrated with Viam.
```

Reviewer 2: Lgtm, maybe without "essentially"?

Author: ```suggestion
In short, any type of hardware can be integrated with Viam.
```

Viam supports a wide variety of sensors, cameras, and other physical hardware, with APIs for each of the following types of hardware:

> {{< cards >}}

Reviewer 2: Not sure where else to put this comment but - given that this is now here, I think we could remove the links to the component API's that appear under "Configure supported hardware"?

Author: We might need to figure out a new place to put them then, because people keep saying they're buried. :/

Reviewer 2: Let's leave them in for now, and discuss with Naomi and Ian.

Jotting down the options I see at the moment:
- Further integrate them with "Configure supported hardware" by adding component filters to the "omni search" box, or splitting up that search box by component. This goes against the idea of directing people to the Registry page in app, which I think is better.
- Having a reference-style page "Component API's" in the left nav

Reviewer: yeah I guess this is one we can talk about between naomi and jessamy and I.  definitely a lot of different opinions here so i don't think we'll come to something async.

Reviewer 3: I don't mind adding a forwarding link to the reference section that goes to the Component APIs page. I don't think it's sufficient. I do think the first point Esha makes is important. If I as a user come to this page wondering "does viam support my arm" it takes quite a bit of text to wade through to get any hint that arms are supported and how to figure out whether my arm is supported. I think something that makes it easily scannable to see "arms are supported"  would be great. Happy to discuss synchronously.

Reviewer 2: Let me know if you'd like me to join for this synchronous discussion :) I would probably lean towards adding a "Browse Component API's" reference-style page so that the left nav is:

* Learn the basics
* Set up a computer or SBC
* Set up an ESP32
* Browse Component API's
* Configure supported hardware
* Integrate other hardware
* Test your machine

"Browse Component API's" would have basically the current Component pages (like [this](https://deploy-preview-4037--viam-docs.netlify.app/operate/reference/components/arm/)), except with the API methods moved above the supported models.
And I'd probably suggest removing the "omni search" box of models from "Configure supported hardware" in favor of linking to the Registry page in app.

I like this because it fits browsing component API's into what feels like an appropriate place in the user flow, while also making it easily find-able as a reference page.
I'm mainly not sure how to square this with the Component API's also existing in the Dev Tools reference.

One related thought is Eliot said he would be fine with moving "Integrate other hardware" out of this section since it's pretty heavyweight.

I know this touches on a lot of aspects, just wanted to get my thoughts down somewhere! If you all want to just keep this in mind for your discussion, that's fine too.

Reviewer 3: Can you tell me what you expect users to be looking for then? What questions do you imagine they have at that point that would be answered by that page?
I think that description would mean that "Browse Component APIs" would not be action orientd. The main goal would be to show/educate the user about Component APIs and I think that should happen while they complete tasks, not separately. Am I seeing that wrong? 

Do you two have ideas for what to do with Integrate other hardware if it's not in this section?

Reviewer 2: Hmm. The tough thing is for some of this more reference content, there are a few different tasks where it's relevant. For example for the Component API's: 

Initially browsing -> "What types of hardware does Viam support? What kinds of API's are there?"
Designing/implementing their machine/solution -> "What were the specific API methods can I call, again?"
Writing a module -> "What API methods do I need to implement?"

It seems reasonable to embed the content in an "early" task a user would do (and Connect Devices is sort of a combination of the first two), and I was thinking we could name it in a way that makes it easy to find for other tasks.

For "Integrate other hardware", it does fit under Build & Integrate. Perhaps it could be its own section at the same level as "Connect Devices" and "Build apps"?

Author: Adding a couple thoughts as I ponder this before Monday:
- I am fine with Integrate Other Hardware staying inside of Connect Devices (though I don't feel strongly and am interested to hear more Esha). Some thoughts:
  - This would force Reference further down the page, and even if we pre-expand it per user feedback, it'd still then be more buried
  - I kinda like how right now Integrate Other Hardware is currently right up next to where we introduce the registry and modular components
  - I'm wondering how we'd name things. If we kept sub-page titles somewhat as they are but in imperative form, e.g. "Create a Hello World module," "Update and manage modules," "Use Micro-RDK modules," they would be Viam-specific. I am not against this since we see a lot of chatbot questions about modules, and it seems that by the point someone is making their own module they might know the word "module" is what they're looking for, *and* "module" is a somewhat-generic term anyway.
- To the point about splitting out browse APIs, they are on the Basics page anyway now. I just put this version together where you can browse by API and it links to each API at the top of the tab--lmk if you have thoughts before Monday and I can mess with other ideas https://github.com/viamrobotics/docs/pull/4087

{{% relatedcard link="/dev/reference/apis/components/encoder/" %}}
{{% relatedcard link="/dev/reference/apis/components/gantry/" %}}
{{% relatedcard link="/dev/reference/apis/components/gripper/" %}}
{{% relatedcard link="/dev/reference/apis/components/input-controller/" %}}
{{% relatedcard link="/dev/reference/apis/components/motor/" %}}
{{% relatedcard link="/dev/reference/apis/components/movement-sensor/" %}}
{{% relatedcard link="/dev/reference/apis/components/power-sensor/" %}}
{{% relatedcard link="/dev/reference/apis/components/sensor/" %}}
{{% relatedcard link="/dev/reference/apis/components/servo/" %}}
{{% relatedcard link="/dev/reference/apis/components/switch/" %}}
{{% relatedcard link="/dev/reference/apis/components/generic/" %}}
{{< /cards >}}

These standardized APIs are implemented by {{< glossary_tooltip term_id="module" text="modules" >}} that provide drivers for specific models of hardware.

Any hardware that is not already supported by a Viam module can be added into Viam's system of modular resources by [creating a new module](/operate/get-started/other-hardware/) that provides a driver for the hardware.

> Viam also supports various software services such as [data capture](/data-ai/capture-data/capture-sync/) and [computer vision](/dev/reference/apis/services/vision/), designed to integrate seamlessly with the hardware driver modules.

Reviewer 2: ```suggestion
```

Suggested moving examples of software services up into the bullets, both to surface them earlier, and because this section is otherwise about hardware.

I also think data capture may eventually become a "system service" rather than an API - it doesn't really make sense to have multiple implementations of it. I suppose people could implement data capture to their own cloud, but the API is pretty specific to the way Viam's cloud is designed.

I think computer vision is a good example, we could either additionally include motion planning or leave it out if that's too much emphasis on motion planning.

> ## How do I get started?

Reviewer 2: As a general comment, is there a way to make the links in this list have the exact text of things in the left nav?

For example, instead of a link on all of "Set up your machine and connect it to the Viam app", the link could be on just "Set up a machine"

Similarly for the other steps

Reviewer 2: The former sounds good - just moving which words the link is on. I was thinking it's a nice visual clue that the left nav and these next steps are the same

Author: Okay sounds like I've already applied this one then cool!

Author: Okay I did tweak a couple, but since set up goes to either of two pages it's not a perfect match and neither is integrate other hardware, though that one contains the phrase. LMK if you think it needs to more strictly match.

Reviewer 2: Thanks for changing those. I was just thinking about this - were we going to move the two setup pages under a common left nav title "Set up a machine"? I still think that's a good idea, since other than the two setup pages, the left nav is meant to be progressed through linearly

Author: (per offline convo leaving as-is for now but open to changing)

1. [Set up your machine and connect it to the Viam app](/operate/get-started/setup/).

1. Next, [configure supported hardware](/operate/get-started/supported-hardware/) and software on your machine in the Viam app.

>    You can [share the configuration across multiple machines](/manage/fleet/reuse-configuration/) to scale quickly.

Reviewer 2: ```suggestion
```

I think it's better for this list to just go through the steps in the left nav on this page; it feels a little distracting ("click off into a different section") to have this here.

Or we could move it below to the "From there, you have many options including" list

Author: okay moving it good call

1. To integrate other hardware or software services that aren't already supported, you can [write your own modules](/operate/get-started/other-hardware/).

1. [Test your machine](/operate/get-started/test/) by using the **TEST** panel of each configured component's config card to, for example, view your camera's stream or turn your motor.

> 1. From there, you have many options including:

Reviewer 2: Similarly not sure where to put this, but I think we should remove the "Configure machine settings" and "Reuse device configuration" left nav links. It feels odd to have some next steps here and some in the left nav, and for them to be different. We could optionally fold those links into the next steps section here

Author: Okay fair point. Naomi may have thoughts since they're to her section and it won't change the main flow here besides the oddness you duly point out.

Reviewer 2: Sounds good - happy to leave in until Naomi returns

Author: @npentrel are you okay with these two being removed from left nav?

Reviewer 3: I worry they won't be discoverable but let's leave them off for now.

> - [Collect data from your machines](/data-ai/capture-data/capture-sync/)

Reviewer 2: ```suggestion
- [Capture data from your machines](/data-ai/capture-data/capture-sync/)
```

- [Create a dataset](/data-ai/ai/create-dataset/) and [train an AI model](/data-ai/ai/train-tflite/)

> - Use an SDK of your choice to [write an app](/operate/control/web-app/) to control your machines

Reviewer 2: ```suggestion
- Use an SDK of your choice to [write an app](/operate/control/web-app/) to interact with your machines
```

Author: Why this change?

Reviewer 2: Hm, it was more because a web app might just pull data from the machine and not actually control it

Author: Okay I'll apply your suggestion

> - [Deploy control logic to run directly on your machines](/manage/software/control-logic/)

Author: ```suggestion
- [Deploy control logic to run directly on your machines](/manage/software/control-logic/)
- [Share the configuration across multiple machines](/manage/fleet/reuse-configuration/).
```
